spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -1136,6 +1140,282 @@ def compile(code)
+          }]
+        $x = MyType('good bye', 'cruel world')
+        notice($x =~ MyType)
+        notice($x =~ MyType['good bye'])
+        notice($x =~ MyType['hello', 'cruel world'])
+        notice($x =~ MyType['good bye', 'cruel world'])
+        PUPPET
+      end
+
+      it 'parameters can be types' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'true', 'true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('good bye', 'cruel world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        notice($x =~ MyType[default, default])
+        notice($x =~ MyType[Enum['hello', 'yo']])
+        PUPPET
+      end
+
+      it 'parameters can be provided using named arguments' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'false', 'true'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => String,
+            p2 => String
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('good bye', 'cruel world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[p1 => 'hello', p2 => 'cruel world'])
+        notice($x =~ MyType[p1 => 'good bye', p2 => 'cruel world'])
+        PUPPET
+      end
+
+      it 'no constraint != default constrant' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('greetings', 'earthlings')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        PUPPET
+      end
+
+      it 'Undef can be used to prevent attribute setting' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'true', 'true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
```

Comment:
if p2 was given `undef` would line 1327 yield `true`? if so, the title is not quite correct as "unset" means "have undef value". If there is a difference it seems a test is missing.

Reply:
It would. And testing undef against undef is tested at lines 1378 - 1380 (where p2 is undef in the instance since it's that's the default).